File: python/src/parse_hcl/utils/serialization/yaml.py
"""
YAML serialization utilities.

Provides custom YAML serialization that matches the TypeScript implementation.
"""

from __future__ import annotations

import json
from typing import Any

PAD = "  "


def to_yaml(value: Any) -> str:
    """
    Converts a value to a YAML-formatted string.

    Handles nested objects, arrays, and scalar values with proper indentation.

    Args:
        value: The value to serialize.

    Returns:
        YAML-formatted string representation.
    """
    return _render(value, 0)


def _render(value: Any, level: int) -> str:
    """
    Recursively renders a value to YAML format.

    Args:
        value: The value to render.
        level: Current indentation level.

    Returns:
        YAML-formatted string for this value.
    """
    if _is_scalar(value):
        return _format_scalar(value)

    if isinstance(value, list):
        if not value:
            return f"{_indent(level)}[]"
        lines = []
        for item in value:
            if _is_scalar(item):
                lines.append(f"{_indent(level)}- {_format_scalar(item)}")
            else:
                rendered = _render(item, level + 1)
                rendered_lines = rendered.split("\n")
                head = rendered_lines[0]
                head_line = head[len(_indent(level + 1)) :] if head.startswith(_indent(level + 1)) else head
                tail = "\n".join(rendered_lines[1:]) if len(rendered_lines) > 1 else ""
                prefix = f"{_indent(level)}- {head_line}"
                lines.append(f"{prefix}\n{tail}" if tail else prefix)
        return "\n".join(lines)

    if isinstance(value, dict):
        if not value:
            return f"{_indent(level)}{{}}"
        lines = []
        for key, val in value.items():
            if _is_scalar(val):
                lines.append(f"{_indent(level)}{key}: {_format_scalar(val)}")
            else:
                rendered = _render(val, level + 1)
                lines.append(f"{_indent(level)}{key}:\n{rendered}")
        return "\n".join(lines)

    # Fallback: use JSON.stringify equivalent for unknown types
    try:
        return f"{_indent(level)}{json.dumps(value)}"
    except (TypeError, ValueError):
        return f"{_indent(level)}{value!r}"


def _format_scalar(value: Any) -> str:
    """
    Formats a scalar value for YAML output.

    Args:
        value: The scalar value to format.

    Returns:
        YAML-formatted string representation.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return f'"{_escape_string(value)}"' if _needs_quoting(value) else value
    if isinstance(value, (int, float)):
        return str(value)
    # Fallback for other types
    try:
        return json.dumps(value)
    except (TypeError, ValueError):
        return repr(value)


def _escape_string(value: str) -> str:
    """
    Escapes special characters in a string for YAML output.

    Args:
        value: The string to escape.

    Returns:
        Escaped string.
    """
    return value.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n").replace("\t", "\\t").replace("\r", "\\r")


def _indent(level: int) -> str:
    """
    Returns the indentation string for a given level.

    Args:
        level: The indentation level.

    Returns:
        Indentation string (spaces).
    """
    return PAD * level


def _is_scalar(value: Any) -> bool:
    """
    Checks if a value is a scalar type.

    Args:
        value: The value to check.

    Returns:
        True if the value is a scalar (None, str, int, float, bool).
    """
    return value is None or isinstance(value, (str, int, float, bool))


def _needs_quoting(value: str) -> bool:
    """
    Checks if a string needs to be quoted in YAML.

    Args:
        value: The string to check.

    Returns:
        True if the string needs quoting.
    """
    if not value:
        return True

    # Check for YAML special characters
    special_chars = set(':#{}[]&*#?|<>=%@`"\'\n\t\r')
    if special_chars.intersection(value):
        return True

    # Check for leading/trailing whitespace
    if value != value.strip():
        return True

    # Check for values that could be interpreted as other types
    lower = value.lower()
    if lower in ("true", "false", "null", "yes", "no", "on", "off"):
        return True

    # Check if it looks like a number
    try:
        float(value)
        return True
    except ValueError:
        pass

    return False

File: typescript/src/utils/serialization/yaml.ts
const PAD = '  ';

export function toYaml(value: unknown): string {
    return render(value, 0);
}

function render(value: unknown, level: number): string {
    if (isScalar(value)) {
        return formatScalar(value);
    }

    if (Array.isArray(value)) {
        if (value.length === 0) {
            return `${indent(level)}[]`;
        }

        return value
            .map((item) => {
                if (isScalar(item)) {
                    return `${indent(level)}- ${formatScalar(item)}`;
                }

                const rendered = render(item, level + 1);
                const lines = rendered.split('\n');
                const first = lines[0].startsWith(indent(level + 1))
                    ? lines[0].slice(indent(level + 1).length)
                    : lines[0];
                const head = `${indent(level)}- ${first}`;
                const tail =
                    lines.length > 1
                        ? lines
                              .slice(1)
                              .map((line) => line)
                              .join('\n')
                        : '';
                return tail ? `${head}\n${tail}` : head;
            })
            .join('\n');
    }

    if (isPlainObject(value)) {
        const entries = Object.entries(value as Record<string, unknown>);
        if (entries.length === 0) {
            return `${indent(level)}{}`;
        }

        return entries
            .map(([key, val]) => {
                if (isScalar(val)) {
                    return `${indent(level)}${key}: ${formatScalar(val)}`;
                }
                const rendered = render(val, level + 1);
                return `${indent(level)}${key}:\n${rendered}`;
            })
            .join('\n');
    }

    return `${indent(level)}${JSON.stringify(value)}`;
}

function formatScalar(value: unknown): string {
    if (value === null || value === undefined) {
        return 'null';
    }
    if (typeof value === 'string') {
        return needsQuoting(value) ? JSON.stringify(value) : value;
    }
    if (typeof value === 'number' || typeof value === 'boolean') {
        return String(value);
    }
    return JSON.stringify(value);
}

function indent(level: number): string {
    return PAD.repeat(level);
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isScalar(value: unknown): boolean {
    return (
        value === null ||
        value === undefined ||
        typeof value === 'string' ||
        typeof value === 'number' ||
        typeof value === 'boolean'
    );
}

function needsQuoting(value: string): boolean {
    return /[:{}\[\],&*#?|<>=%@`]/.test(value) || value.includes('"') || value.includes("'") || value.includes('\n');
}
